calculate_daily_summary.py

```diff
@@ -55,6 +60,48 @@
     ORDER BY 1;
 """
 
+# For the monthly device activity summary,
+# we maintain a table giving unique (uid, device_id, service)
+# values seen in the last 28 days
+
+Q_MONTHLY_ACTIVITY_CREATE_TABLE = """
+    CREATE TABLE IF NOT EXISTS monthly_unique_activity (
+      day DATE NOT NULL SORTKEY ENCODE lzo,
+      uid VARCHAR(64) NOT NULL DISTKEY ENCODE lzo,
+      device_id VARCHAR(32) NOT NULL ENCODE lzo,
+      service VARCHAR(40) ENCODE lzo,
+      ua_browser VARCHAR(40) ENCODE lzo,
+      ua_version VARCHAR(40) ENCODE lzo,
+      ua_os VARCHAR(40) ENCODE lzo
+    );
+"""
+
+Q_MONTHLY_ACTIVITY_CLEAR = """
+    DELETE FROM monthly_unique_activity
+    WHERE day >= '{day_from}'::DATE
+    AND day <= '{day_until}'::DATE;
+"""
+
+Q_MONTHLY_ACTIVITY_SUMMARIZE = """
+    INSERT INTO monthly_unique_activity
+      (day, uid, device_id, service, ua_browser, ua_version, ua_os)
+    SELECT DISTINCT
+      days.day, past.uid, past.device_id, past.service,
+      past.ua_browser, past.ua_version, past.ua_os
+    FROM (
+      SELECT DISTINCT day FROM daily_activity_per_device
+      WHERE day >= '{day_from}'::DATE AND day <= '{day_until}'::DATE
+      ORDER BY 1
+    ) AS days,
+    INNER JOIN daily_activity_per_device AS past
+      ON past.day <= days.day
+      AND past.day > (days.day - '28 days'::INTERVAL)
+      -- This is not strictly necessary,
+      -- but might help speed the query up.
+      AND past.day > ('{day_from}'::DATE - '28 days'::INTERVAL)
+    ORDER BY 1;
+"""
```

Comment:
This is subtle, and I'm not sure the naming here is very effective at conveying what it means, but I haven't yet come up with something better.

The table is named "monthly_unique_activity", but it contains data on a daily basis.  There's a record in this table for ($DAY, ....blah....) IFF there's a record in the daily-activity table for any of the 28 days preceeding $DAY.  I guess we could call it "daily_summary_of_unique_activity_over_previous_28_days" but that seems a little unwieldy :-)

Actually "unique_activity_in_previous_month" is not *too* bad, come to think of it...